Emit function definitions in check, for all specifics seen.

alinas opened the pull request:
Emitting definitions in check. This resolves the crash in lowering which necessitated definitions be emitted.
Some of the test changes need further review.

alinas committed:
Emit function definitions in check, for all specifics seen.

alinas commented:
Created PR to get feedback. Marked as draft because I want to review the tests further.

geoffromer commented on toolchain/check/check_unit.cpp:
```diff
@@ -473,33 +476,38 @@ auto CheckUnit::CheckRequiredDefinitions() -> void {
     }
   }
+
+  auto handle_specific_function = [&](SemIRLoc loc,
```
This looks like it only has one callsite. Could it be inlined there, or is there some subtle reason that doesn't work?

geoffromer commented on toolchain/check/context.h:
```diff
@@ -334,6 +339,11 @@ class Context {
   // end of the current source file.
   llvm::SmallVector<SemIR::InstId> definitions_required_;
 
+  // Functions that should have definitions by the end of the current source
```
Should this be limited to functions, or can it be applicable to classes etc. as well? On the flip side, is it applicable to all functions, or only to generic functions?

What's the meaning of the `SemIRLoc`?

alinas replied:
AFAICT yes this is restricted to functions. This is populated when a Call to a function is found in a generic context to another generic function, and a definition needs to be later added to this. This is only for generic functions.

The SemIRLoc is the location for where the specific is evaluated as a constant, used for diagnostics purposes.

geoffromer replied:
I'm asking about this because it's confusing to have both `definitions_required` and `function_definitions_required`, especially when some things that require definitions are not in `definitions_required`, and some functions that require definitions are not in `function_definitions_required`. It seems like the names and comments need to be updated.

It seems like `definitions_required_` is really tracking entities where a definition is required as soon as it's declared (and so it only needs to record the declaration), whereas `function_definitions_required_` is tracking specifics where a definition is required only if it's actually used (so it needs to record the entity that needs a definition, _and_ it needs to record the point of use, for diagnostics). So maybe they could be called `definitions_required_by_decl` and `definitions_required_by_use`.

alinas replied:
I like those names more, you're right, those were confusing. I'll make that change.

geoffromer commented on toolchain/check/testdata/builtin_conversions/no_prelude/convert_class_type_to_generic_facet_value.carbon:
```diff
@@ -433,8 +433,6 @@ fn G() {
 // CHECK:STDOUT:   %U.loc17_32.2 => constants.%Generic.facet
 // CHECK:STDOUT:   %U.patt.loc17_32.2 => constants.%Generic.facet
-// CHECK:STDOUT:
-// CHECK:STDOUT: !definition:
```
Are these changes expected? I can sort of see how this PR could lead to some specifics gaining definitions, but it's harder to understand why some lose their definitions (or am I misunderstanding what `!definition:` signifies?)

zygoloid replied:
This is happening for specifics whose argument list contains an error. In that case we don't reach the `EvalConstantInst` call, so we don't attempt to create a definition any more.

alinas committed:
Merge branch 'trunk' into emit_definition

alinas committed:
Update tests.

alinas committed:
Inline lambda

alinas committed:
Inline check for self_type in callee.

alinas committed:
Update names of definitions required.

This better reflects the reason the definitons are expected to exist by
end of file.

zygoloid commented on toolchain/check/check_unit.cpp:
```diff
@@ -474,33 +477,32 @@ auto CheckUnit::CheckRequiredDefinitions() -> void {
   }
+
+  // Note that more required definitions can be added during this loop.
+  // NOLINTNEXTLINE
```
It'd be helpful to list the name of the lint being disabled here: `NOLINTNEXTLINE(modernize-loop-convert)` (I'm assuming that's the check that's firing here).

zygoloid commented on toolchain/check/eval_inst.cpp:
```diff
@@ -348,13 +349,30 @@ auto EvalConstantInst(Context& context, SemIRLoc loc,
   auto specific_id = MakeSpecific(context, loc, generic_id, args);
 
-  // TODO: Add the new `SpecificFunction` to definitions_required.
+  if (!SemIR::GetCalleeFunction(context.sem_ir(), inst.callee_id)
+           .self_type_id.has_value()) {
```
I think we don't need this check in this case -- we just fetched the function from an impl witness table, so it can't be a function declaration from an interface.

alinas committed:
Address comments.

geoffromer commented on toolchain/check/check_unit.cpp:
```diff
@@ -444,8 +444,9 @@ auto CheckUnit::CheckRequiredDefinitions() -> void {
   // Note that more required definitions can be added during this loop.
-  for (size_t i = 0; i != context_.definitions_required().size(); ++i) {
-    SemIR::InstId decl_inst_id = context_.definitions_required()[i];
+  // NOLINTNEXTLINE (modernize-loop-convert)
```
```suggestion
  // NOLINTNEXTLINE(modernize-loop-convert)
```

If there's a space between them, `(modernize-loop-convert)` gets interpreted as a free-text comment, and so this would silence all clang-tidy warnings. See [here](https://clang.llvm.org/extra/clang-tidy/#clang-tidy-nolint).

(Idle thought: making `definitions_required_by_decl` a queue and looping until it's empty might make this code clearer, and would avoid the need to suppress this warning. I don't know if it's worth the trouble, though.)

geoffromer commented on toolchain/check/check_unit.cpp:
```diff
@@ -484,12 +485,11 @@ auto CheckUnit::CheckRequiredDefinitions() -> void {
-  // NOLINTNEXTLINE
-  for (size_t i = 0; i != context_.function_definitions_required().size();
-       ++i) {
+  // NOLINTNEXTLINE (modernize-loop-convert)
```
```suggestion
  // NOLINTNEXTLINE(modernize-loop-convert)
```

alinas committed:
Address comment

alinas committed:
Merge branch 'trunk' into emit_definition

alinas committed:
Merge branch 'trunk' into emit_definition
